jerry-core/lit/lit-magic-strings.c

```diff
@@ -264,22 +283,57 @@ lit_magic_string_ex_id_t
 lit_is_ex_utf8_string_magic (const lit_utf8_byte_t *string_p, /**< utf-8 string */
                              lit_utf8_size_t string_size) /**< string size in bytes */
 {
-  /* TODO: Improve performance of search */
+  const uint32_t magic_string_ex_count = lit_get_magic_string_ex_count ();
 
-  for (lit_magic_string_ex_id_t id = (lit_magic_string_ex_id_t) 0;
-       id < JERRY_CONTEXT (lit_magic_string_ex_count);
-       id = (lit_magic_string_ex_id_t) (id + 1))
+  if (magic_string_ex_count == 0
+      || string_size > lit_get_magic_string_ex_size (magic_string_ex_count - 1))
   {
-    if (string_size == lit_get_magic_string_ex_size (id))
+    return (lit_magic_string_ex_id_t) magic_string_ex_count;
+  }
+
+  lit_magic_string_ex_id_t first = 0;
+  lit_magic_string_ex_id_t last = (lit_magic_string_ex_id_t) magic_string_ex_count;
+  bool found_size_range = false;
+
+  while (first < last)
+  {
+    const lit_magic_string_ex_id_t middle = (first + last) / 2;
+    const lit_utf8_byte_t *ext_string_p = lit_get_magic_string_ex_utf8 (middle);
+    const lit_utf8_size_t ext_string_size = lit_get_magic_string_ex_size (middle);
+
+    if (string_size == ext_string_size)
     {
-      if (memcmp (string_p, lit_get_magic_string_ex_utf8 (id), string_size) == 0)
+      found_size_range = true;
+      const int string_compare = memcmp (ext_string_p, string_p, string_size);
+
+      if (string_compare == 0)
+      {
+        return middle;
+      }
+      else if (string_compare < 0)
       {
-        return id;
+        first = middle + 1;
       }
+      else
+      {
+        last = middle;
+      }
+    }
+    else if (found_size_range)
+    {
+      break;
```

Comment:
Are you sure this is correct?

a
b
c
d
e
ff
gg

Search e.

First, you find d, and found_size_range is set to 1. Then you find ff, and the loop aborts. If I understand correctly.